cf-oscore/src/main/java/org/eclipse/californium/oscore/ContextRederivation.java

```diff
@@ -18,86 +18,376 @@
 			LOGGER.error(ErrorDescriptions.CONTEXT_REGENERATION_FAILED);
 			throw new CoapOSException(ErrorDescriptions.CONTEXT_REGENERATION_FAILED, ResponseCode.BAD_REQUEST);
 		}
 	}
-	
+
+	/* Client side related methods below */
+
 	/**
-	 * Perform re-derivation of contexts as detailed in Appendix B.2.
-	 * Essentially it uses a message exchange together with the Context ID
-	 * field in the OSCORE option to securely generate a new shared context.
+	 * This method will be triggered by the client to send a request that
+	 * initiates the context re-derivation procedure.
 	 * 
-	 * @throws IOException 
-	 * @throws ConnectorException 
-	 * @throws OSException 
+	 * @throws IOException
+	 * @throws ConnectorException
+	 * @throws OSException
 	 */
-	private static void rederive(OSCoreCtxDB db, String uri) throws ConnectorException, IOException, OSException {
-		//Generate a random 8 byte Context ID
-		SecureRandom random = new SecureRandom();
-		byte[] newContextId = new byte[8];
-		random.nextBytes(newContextId);
-		
-		//Retrieve the context for the target URI
-		OSCoreCtx oldCtx = db.getContext(uri);
-		
-		//Create new context with the generated Context ID
-		OSCoreCtx ctx = new OSCoreCtx(oldCtx.getMasterSecret(), true, oldCtx.getAlg(), oldCtx.getSenderId(),
-				oldCtx.getRecipientId(), oldCtx.getKdf(), oldCtx.getRecipientReplaySize(), oldCtx.getSalt(), newContextId);
-		ctx.setIncludeContextId(true);
-		db.addContext(uri, ctx);
-		
-		//Now send request using the new context
+	private static void initiateRequest(OSCoreCtxDB db, String uri) throws ConnectorException, IOException, OSException {
+		// Retrieve the context for the target URI
+		OSCoreCtx ctx = db.getContext(uri);
+
+		// Generate a random Context ID (ID1)
+		byte[] contextID1 = Bytes.createBytes(random, SEGMENT_LENGTH);
+
+		// Create new context with the generated Context ID
+		OSCoreCtx newCtx = rederiveWithContextID(ctx, contextID1);
+		newCtx.setIncludeContextId(true);
+		newCtx.setContextRederivationPhase(ContextRederivation.PHASE.CLIENT_PHASE_1);
+		db.addContext(uri, newCtx);
+
+		// Now a send request using the new context
 		String resource = "/rederive"; //Dummy resource to access for context re-derivation
 		String URI = uri + resource;
-		System.out.println(URI);
 		
 		CoapClient c = new CoapClient(URI);
 		Request r = new Request(Code.GET);
 		r.getOptions().setOscore(Bytes.EMPTY);
-		System.out.println(Utils.prettyPrint(r));
-		
-		CoapResponse resp = null;
-		resp = c.advanced(r);
-		System.out.println(Utils.prettyPrint(resp));
-		c.shutdown();
+
+		CoapHandler handler = new CoapHandler() {
+
+			@Override
+			public void onLoad(CoapResponse response) {
+				LOGGER.info("Successfully received context re-derivation message");
+			}
+
+			@Override
+			public void onError() {
+				LOGGER.info("Error while receiving context re-derivation message");
+			}
+		};
+
+		c.advanced(handler, r);
+
+		try {
+			Thread.sleep(1000);
```

Comment:
In close to any case: that indicates mainly a threading issue.
With CoapHandler  the implementation pretends to be asyn, with sleep it get pseudo sync again.

Reply:
Done in commit 89447884a6cbf853d0e3e34b0a73e4cad4a24f1f